jsonrpc/websocket.go

```diff
@@ -54,9 +54,19 @@ func (ws *Websocket) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 	// TODO include connection information, such as the remote address, in the logs.
 
-	wsc := newWebsocketConn(conn, ws.rpc, ws.connParams, ws.listener)
+	wsc := newWebsocketConn(r.Context(), conn, ws.connParams, ws.listener)
 
-	err = wsc.ReadWriteLoop(r.Context())
+	for {
+		// Handle will read from the connection once.
+		var resp []byte
+		resp, err = ws.rpc.HandleReader(wsc.ctx, wsc)
+		if err != nil {
+			break
+		}
+		if _, err = wsc.Write(resp); err != nil {
+			break
+		}
```

Comment:
We'll need to make sure subscriptions don't write a notification before we respond to the initial request.